### ISSUE TITLE
CRASH: Unable to open a realm at path

### ISSUE THREAD

**Issue by jln19** [realm/realm-cocoa#7129]
I can find many similar issues here and on StackOverflow, but there don't seem to be any solutions. This is a fatal issue for our app so I'm hoping that surfacing it again can lead to some insights for myself and others.

Thread 1: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=2 "Unable to open a realm at path '/path/to/myfile.realm': Invalid top array (ref: 3024, size: 11)

I have not made any changes to any Realm objects.

## Expected Results
App doesn't crash when opening.

## Actual Results
App crashes when opening.

## Steps for others to Reproduce
Install live version of the app from App Store. Then run the app from Xcode or other testing service such as TestFlight.

If I delete the app from the device, and do a fresh build from Xcode, the app doesn't crash. It only happens when upgrading from a previous version.

## Code Sample
```
private lazy var realm: Realm = {
        return try! Realm()
    }()
```

## Version of Realm and Tooling
Realm framework version: ?
Latest ( pod 'RealmSwift' )
Xcode version: ?
12.2
iOS/OSX version: ?
14.2

**Comment by jln19** [comment]
The only way to get this not to crash is to create a new config.fileURL on every update:

```
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        if let a = appVersion {
            config.fileURL = config.fileURL!.deletingLastPathComponent().appendingPathComponent("file\(a.replacingOccurrences(of: ".", with: "")).realm")
        } else {
            config.fileURL = config.fileURL!.deletingLastPathComponent().appendingPathComponent("file2.realm")
        }
```

Obviously this is not ideal and I'm sure there are ramifications for doing this. But the only other option is for all of our users to experience a crash every time they update and be completely unable to use the app unless they delete it and re-install from the app store.

**Comment by DominicFrei** [comment]
> Thread 1: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=2 "Unable to open a realm at path '/path/to/myfile.realm': Invalid top array (ref: 3024, size: 11)

@jln19 Thanks for the report. I'd like to understand your setup a bit better. The above error message indicated that the path (alongside with maybe other properties) was changed and then set as `defaultConfiguration` which would be used when you open the realm the way you did.

Can you show us the rest of the relevant realm code?

Unfortunately I cannot reproduce your issue with `Install live version of the app from App Store. Then run the app from Xcode or other testing service such as TestFlight.` since I don't have the app or the code.
It sounds like the App Store version and the one you are then trying to run locally do not have the same path.

**Comment by jln19** [comment]
Hi @DominicFrei,  I haven't changed the path. Here is the full Realm code:

```
func configure() {
        var config = Realm.Configuration(

            // Set the new schema version. This must be greater than the previously used
            // version (if you've never set a schema version before, the version is 0).
            schemaVersion: 2,

            // Set the block which will be called automatically when opening a Realm with
            // a schema version lower than the one set above
            migrationBlock: { migration, oldSchemaVersion in
                // We haven’t migrated anything yet, so oldSchemaVersion == 0
                if (oldSchemaVersion < 2) {
                    // Nothing to do!
                    // Realm will automatically detect new properties and removed properties
                    // And will update the schema on disk automatically
                }
            }

        )

        config.fileURL = config.fileURL!.deletingLastPathComponent().appendingPathComponent("store1.realm")

        // Tell Realm to use this new configuration object for the default Realm
        Realm.Configuration.defaultConfiguration = config
    }
```

**Comment by DominicFrei** [comment]
> Thread 1: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=io.realm Code=2 "Unable to open a realm at path '/path/to/myfile.realm': Invalid top array (ref: 3024, size: 11)

I've looked into this further and found the error message originating in https://github.com/realm/realm-core/blob/master/src/realm/group.cpp#L677.

This is an `InvalidDatabase` error which explains why your workaround in
> ```
> let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
> if let a = appVersion {
>     config.fileURL = config.fileURL!.deletingLastPathComponent()
>             .appendingPathComponent("file\(a.replacingOccurrences(of: ".", with: "")).realm")
> } else {
>     config.fileURL = config.fileURL!.deletingLastPathComponent().appendingPathComponent("file2.realm")
> }
> ```
works.

I'll be talking to our core colleagues to figure out if there is a way to recover from this.

Another question that came up when re-reading your explanation. @jln19 
> The only way to get this not to crash is to create a new config.fileURL on every update

This happens on every single update (migration?) you make? That would hint to something general being wrong.

**Comment by DominicFrei** [comment]
@jln19 The exception originates in https://github.com/realm/realm-core/blame/master/src/realm/group.cpp#L684-L685 actually.

It looks like the realm file size is 11, which was added in https://github.com/realm/realm-core/commit/645ea73c5886cfaea2ca0fcd9da0cf6d6e34b11a#.
Realm files generated with Realm Core version 6.0.0+ (which has been released with Realm Cocoa 5.0.0) do have size 11. When you try to open those files with a version prior to Realm Core 6.0.0 / Realm Cocoa 5.0.0 the above error can occur.

While this is just one possible cause I'd like to investigate further by understanding your versioning.

Which version of the Realm are you using at the moment? Which version did you use to generate the file in question? Was there any downgrading at any time?

Also, if you could point me to the other issues you mentioned I would be really glad. Maybe there is some additional information in there so that I can tackle this issue.

**Comment by leemaguire** [comment]
@jln19 would it be possible to inspect the app's sandbox after the upgrade and see if the Realm file exists?
